File: 백준/Gold/9489. 사촌/사촌.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        // 빠른 입력을 위해 BufferedReader 사용
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());

            // 종료 조건
            if (n == 0 && k == 0) break;

            int[] nodes = new int[n];
            int k_idx = 0;
            
            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) {
                nodes[i] = Integer.parseInt(st.nextToken());
                if (nodes[i] == k) {
                    k_idx = i; // k의 위치(인덱스) 미리 저장
                }
            }

            int[] parent_idx = new int[n];
            Arrays.fill(parent_idx, -1); // 부모 배열을 -1로 초기화
            
            int cur_parent_idx = -1;
            
            // 1. 부모 인덱스 설정
            for (int i = 1; i < n; i++) {
                if (nodes[i] != nodes[i - 1] + 1) {
                    cur_parent_idx++;
                }
                parent_idx[i] = cur_parent_idx;
            }

            // 2. 사촌 찾기
            int k_parent = parent_idx[k_idx];
            int ans = 0;
            
            // k가 루트가 아니고(-1), 할아버지가 존재할 때(parent_idx[k_parent] != -1)만 실행
            if (k_parent != -1 && parent_idx[k_parent] != -1) {
                int k_grandparent = parent_idx[k_parent];
                
                for (int i = 1; i < n; i++) {
                    // 부모는 다르고, 할아버지는 같은 경우
                    if (parent_idx[i] != k_parent) {
                        // parent_idx[i]가 -1이면 할아버지를 찾을 때 에러가 날 수 있으므로 체크
                        if (parent_idx[i] != -1 && parent_idx[parent_idx[i]] == k_grandparent) {
                            ans++;
                        }
                    }
                }
            }
            
            System.out.println(ans);
        }
    }
}

File: 백준/Gold/9489. 사촌/사촌.py
# 사촌

# k 의 부모와 다르고, 부모의 부모가 일치하면 카운트

while True:
    n, k = map(int,input().split())
    
    if n == 0 and k == 0: # 종료 조건
        break
    
    nodes = list(map(int,input().split()))
    
    parent = [-1] * n # 각 노드의 부모 노드의 인덱스를 저장할 배열

    
    cur_parent_idx = -1 # 현재 부모의 인덱스
    # cur_parent_idx = 0

    # parent[0] = -1
    
    for i in range(1, n): # 모든 노드에 부모 만들어주기
        if nodes[i] == nodes[i-1] + 1: # 값이 연속인가? (같은 부모라는 뜻)
            # 같은 부모의 자식들은 값이 연속된 증가로 주어짐
            parent[i] = cur_parent_idx
        else: # 부모가 달라짐
            cur_parent_idx += 1
            parent[i] = cur_parent_idx
    
    k_idx = nodes.index(k)
    k_parent_idx = parent[k_idx]
    
    ans = 0

    # k의 부모가 존재하고, 그 부모의 부모(조부모)도 존재해야 사촌이 가능
    if k_parent_idx != -1 and parent[k_parent_idx] != -1:
        k_grandparent_idx = parent[k_parent_idx] # 부모의 부모
        for i in range(1,n):
            if parent[i] != k_parent_idx and parent[parent[i]] == k_grandparent_idx:
                ans += 1
    else:
        ans = 0

    print(ans)